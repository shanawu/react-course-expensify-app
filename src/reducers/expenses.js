// Expense Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    // add expense
    case "ADD_EXPENSE":
      return [...state, action.expense]; // spread operator to add the newly added expense without modifying the original value
    // remove expense
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id); // state.filter((expense) => expense.id !== action.expense.id)
    // edit expense
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          // finding if there is a match in the expenses
          return {
            ...expense, // spread the expense
            ...action.updates, // spread the updates
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
