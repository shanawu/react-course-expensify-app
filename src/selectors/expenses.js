import moment from "moment";
// Get Filtered Expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch =
        typeof text === undefined ||
        expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
      // typeof text === undefined || expense.description.toLowerCase().includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1; // 1: b come first, -1: a come first
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
