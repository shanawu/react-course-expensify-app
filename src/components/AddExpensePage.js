import React from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

const AddExpensePage = (props) => {
  const navigate = useNavigate();
  const handleExpenseSubmit = (expense) => {
    console.log("Submitting expense:", expense); // Add this line

    props.dispatch(addExpense(expense));
    navigate("/");
  };

  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={handleExpenseSubmit} />
    </div>
  );
};

export default connect()(AddExpensePage);
