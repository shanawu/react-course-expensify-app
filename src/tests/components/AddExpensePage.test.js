import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "../../store/configureStore";
import AddExpensePage from "../../components/AddExpensePage";
import expenses from "../fixtures/expenses";

test("should render Add Expense page correctly", () => {
  const store = configureStore();
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <AddExpensePage />
      </MemoryRouter>
    </Provider>
  );
  expect(container).toMatchSnapshot();
});

test("should handle expense form submission correctly", () => {
  const store = configureStore();
  const { container, getByTestId } = render(
    <Provider store={store}>
      <MemoryRouter>
        <AddExpensePage />
      </MemoryRouter>
    </Provider>
  );

  // Fill in the form fields with test data
  const descriptionInput = getByTestId("description-input");
  const amountInput = getByTestId("amount-input");
  const noteTextarea = getByTestId("note-textarea");

  fireEvent.change(descriptionInput, {
    target: { value: expenses[0].description },
  });
  fireEvent.change(amountInput, { target: { value: expenses[0].amount } });
  fireEvent.change(noteTextarea, {
    target: { value: expenses[0].note },
  });

  // Submit the form
  fireEvent.submit(getByTestId("expense-form"));

  // After the form is submitted, the `handleExpenseSubmit` function should be called with the correct data
  expect(store.getState().expenses).toEqual([
    {
      id: expect.any(String), // The id will be auto-generated, so we just check its presence here
      description: expenses[0].description,
      amount: expenses[0].amount * 100,
      note: expenses[0].note,
      createdAt: expect.any(Number), // The createdAt timestamp will be generated during form submission
    },
  ]);
});
