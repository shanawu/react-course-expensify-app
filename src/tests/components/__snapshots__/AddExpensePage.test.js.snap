// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Add Expense page correctly 1`] = `
<div>
  <div>
    <h1>
      Add Expense
    </h1>
    <div>
      <form
        data-testid="expense-form"
      >
        <input
          data-testid="description-input"
          placeholder="Description"
          type="text"
          value=""
        />
        <input
          data-testid="amount-input"
          placeholder="Amount"
          type="text"
          value=""
        />
        <div
          class="SingleDatePicker SingleDatePicker_1"
        >
          <div>
            <div
              class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
            >
              <div
                class="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-date"
                  aria-label="Date"
                  autocomplete="off"
                  class="DateInput_input DateInput_input_1"
                  id="date"
                  name="date"
                  placeholder="Date"
                  type="text"
                  value="07/23/2023"
                />
                <p
                  class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-date"
                >
                  Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
            </div>
          </div>
        </div>
        <textarea
          data-testid="note-textarea"
          placeholder="Add a note for your expense (optional)"
        />
        <button
          data-testid="expense-form-submit"
        >
          Add a expense
        </button>
      </form>
    </div>
  </div>
</div>
`;
